### ISSUE TITLE
Improve usability of prompt themes using asynchronicity 

### ISSUE THREAD

**Issue by sorin-ionescu** [sorin-ionescu/prezto#796]
Zsh supports process forking, and through it, prompts, at least `RPROMPT`, can be [updated asynchroniouslly](http://www.anishathalye.com/2015/02/07/an-asynchronous-shell-prompt/).

By moving computationally expensive calculations, for example, Git repository status, to `RPROMPT`, and calculating it asynchroniouslly, user experience improves since `PROMPT` is instantly available for command entering.

The same can be done for `PROMPT`, but it requires knowing how many characters the result of a computationally expensive operation takes and using placeholder text to later replace to prevent the prompt from jumping to the right.

![Async Prompt](http://www.anishathalye.com/media/2015/02/07/demo.gif)

**Comment by leoj3n** [comment]
Looks way faster :snail: Git repos do add a big delay.

**Comment by PaBLoX-CL** [comment]
This is an interesting approach. Even in some not too big repos, the delay is very noticeable. :+1:

**Comment by sorin-ionescu** [comment]
@PaBLoX-CL I have toyed with it in 50f4895e72d6106839688bc24c4f20404542e955, but it's not working. The SIG USR1 trap is being deleted, and I do not know why.

**Comment by sorin-ionescu** [comment]
I have figured it out. Traps are contextual and are reset.

**Comment by sorin-ionescu** [comment]
I have got it to work in emacs mode.

In vi mode, Git status disappears upon entering command mode.

**Comment by sorin-ionescu** [comment]
I fixed it for vi as well.

`typeset -p` won't print the original declaration. I had to use string substitution to change `typeset -A` to `typeset -Ag` in `$git_info` serialisation. Is there a cleaner way?

**Comment by 2vek** [comment]
:+1: This will be awesome improvement.

Sometimes It do annoy me to wait for tab completions when `cd` to git repo or opening the terminal for first time after system boot.

**Comment by lazywei** [comment]
How is this feature going? I'm looking forward to this :D

**Comment by sorin-ionescu** [comment]
I've been using it for the past 5 weeks. It works very well. The only
annoyance is that you have to move all slow code to the right prompt; but,
it is worth it since the usability improvements are significant.

I've only changed my theme. I might change the PeepCode theme as well since
it is an example of git-info. I am not touching other themes.

**Comment by lazywei** [comment]
I guess it hasn't been merged into master, right? cause I have the latest prezto with sorin prompt, but it seems I don't have such feature yet.
Thanks.

**Comment by sorin-ionescu** [comment]
It has not been merged.

**Comment by michaelpj** [comment]
Any chance of this making it to master sometime soon? It looks excellent!

**Comment by lazywei** [comment]
same here. would like to have this in master :)

**Comment by sloria** [comment]
+1. Let us know if any help is needed to get the branch merged.

**Comment by sorin-ionescu** [comment]
I've done this just for my theme.

**Comment by sorin-ionescu** [comment]
This was done in a1dea6ae43f5a1f5e5dfabac49766bda67983ea0; a few aesthetic changes followed.

**Comment by raine** [comment]
Unfortunately I'm getting weird blinking in tmux unless the pane is full of content:

![prezto](https://cloud.githubusercontent.com/assets/11027/7793944/21cf791a-02ce-11e5-996d-0f1dd564b743.gif)

**Comment by sorin-ionescu** [comment]
Weird. I have not noticed that outside of tmux. What about the original theme? http://www.anishathalye.com/2015/02/07/an-asynchronous-shell-prompt/

**Comment by jck** [comment]
I'm not experiencing the tmux issue. I'm running tmux 2.0 and zsh 5.0.7

**Comment by gmmeyer** [comment]
![optimised](https://cloud.githubusercontent.com/assets/3253673/7818914/7bca1216-03ac-11e5-8826-f1a48587cafc.gif)

This is my error, it's slightly different, though. It's not erratic cursor behavior, it's certainly unwanted, though.

**Comment by mafredri** [comment]
Regarding the blinking issue: I experimented with `zle reset-prompt` when rewriting the async functionality of `pure`. In my experience it caused flickering of the prompt. Reason remains unknown. Due to this `pure` still relies fully on terminal codes to reset the prompt, e.g. storing cursor position, rewriting, erasing, etc..

**Comment by sorin-ionescu** [comment]
If it flickers, it should only flicker once when it's updated; it should not cause a blinking cursor.

**Comment by PaBLoX-CL** [comment]
@raine, sorry for the hijack... which theme are you using?

**Comment by raine** [comment]
@PaBLoX-CL it's the default `sorin` theme, but I changed the PWD color back to a more pleasing one.

**Comment by wikimatze** [comment]
I have similar issues as @raine describes. The "flickering" only occurs in tmux session. I'm using xfce4-terminal 0.6.3git-f7c72e5 and tmux 2.2. Here is the error I get:


![blinking_tmux](https://cloud.githubusercontent.com/assets/264708/12245132/35b399fe-b8a7-11e5-9e57-22c571a5c185.gif)

I tried a lot to experiment with my  `.zpretzorc` file but finally found the source of error in my `tmux.conf` file. I had to remove the new 'dim' feature from it:


```sh
set-window-option -g window-style 'bg=#181818'
set-window-option -g window-active-style 'bg=black'
set-window-option -g pane-active-border-style ''
```

After that, everything was working fine. Hope, this can help some people.
